airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeRecordConsumer.java

```diff
@@ -106,28 +108,43 @@ private static void executeWriteQuery(Supplier<Connection> connectionFactory,
                                         CloseableQueue<byte[]> writeBuffer,
                                         String tmpTableName) {
     try (Connection conn = connectionFactory.get()) {
-      conn.setAutoCommit(false);
+      final List<AirbyteRecordMessage> records = Lists.newArrayList();
+      for (int i = 0; i < batchSize; i++) {
+        final byte[] record = writeBuffer.poll();
+        if (record == null) {
+          break;
+        }
+        final AirbyteRecordMessage message = Jsons.deserialize(new String(record, Charsets.UTF_8), AirbyteRecordMessage.class);
+        records.add(message);
+      }
+
+      LOGGER.info("max size of batch: {}", batchSize);
+      LOGGER.info("actual size of batch: {}", records.size());
+
+      if (records.isEmpty()) {
+        return;
+      }
 
-      try (PreparedStatement statement = conn.prepareStatement(
-          String.format("INSERT INTO \"%s\" (\"ab_id\", \"%s\", \"emitted_at\") SELECT ?, parse_json(?), ?", tmpTableName,
-              SnowflakeDestination.COLUMN_NAME))) {
+      final StringBuilder sql = new StringBuilder().append(String.format(
+          "INSERT INTO \"%s\" (\"ab_id\", \"%s\", \"emitted_at\") SELECT column1, parse_json(column2), column3 FROM VALUES\n",
+          tmpTableName,
+          SnowflakeDestination.COLUMN_NAME));
 
-        for (int i = 0; i < batchSize; i++) {
-          final byte[] record = writeBuffer.poll();
-          if (record == null) {
-            break;
-          }
-          final AirbyteRecordMessage message = Jsons.deserialize(new String(record, Charsets.UTF_8), AirbyteRecordMessage.class);
+      records.forEach(r -> sql.append("(?, ?, ?),\n"));
```

Comment:
a bit of a song and dance here to be able to still used the `PreparedStatement` to bind values. procedure:
1. loop through records once to create the sql string
1. create `PreparedStatement` using that string
1. loop through the records again to bind the values to the string using the `PreparedStatement` iface

Can't do one loop because the `PreparedStatement` needs to have the full SQL string at the time it is created. The values are bound after.

Not loving this, but I think with a comment or two explaining what's going on it should serve.

Reply:
Not loving this either but I see why it's here. Could you add the comment?

Reply:
yes!